OpenRA.Mods.Common/Activities/Air/FlyForward.cs

```diff
@@ -36,11 +49,15 @@ public override bool Tick(Actor self)
 				return true;
 			}
 
-			if (IsCanceling || remainingTicks-- == 0)
+			// Having flyTicks < 0 is valid and means the actor flies until this activity is canceled
+			if (IsCanceling || (flyTicks > 0 && ticks++ >= flyTicks) || (flyTicks == 0 && remainingDistance <= 0))
 				return true;
 
-			Fly.FlyTick(self, aircraft, aircraft.Facing, cruiseAltitude);
+			// FlyTick moves the aircraft while FlyStep calculates how far we are moving
+			if (remainingDistance != 0)
+				remainingDistance -= aircraft.FlyStep(aircraft.Facing).HorizontalLength;
```

Comment:
This will not give the correct distance when the aircraft gets repulsed along the way. This could be sidestepped by comparing the actual current position with the position on the previous tick instead.

Reply:
Depends on what we want to do, let the aircraft *fly forward* or *cover a distance*?

Reply:
If the point is not to cover a set distance, then we shouldn't allow constructing with a distance argument in the first place. 

If the point is just to fly forward for some number of ticks&mdash;even when repulsion forces may in the extreme case make the aircraft barely move at all&mdash;then just define a number of ticks and don't bother.

Reply:
Measuring the actual distance is an easy fix - rename `remainingDistance` to `distance` and make it readonly, add an `OnFirstRun` that sets a `WPos distanceFrom` variable, then check `(self.CenterPosition - distanceFrom).HorizontalLengthSquared >= distance.LengthSquared`

Reply:
That would be measuring a different kind of distance: The cartesian distance from the starting point instead of the distance travelled along the trajectory. Tbh, I'm not sure which one would be better, but that wasn't the issue.

The issue is _not_ that the trajectory could be bent by repulsion&mdash;I'd assume from the code that abc intended to measure the bent trajectory; the issue is that because `FlyStep` doesn't take repulsion into account, this code doesn't _actually_ measure the distance travelled along the trajectory either!

Reply:
The code that uses `FlyForwards` isn't smart enough to handle repulsion either, so i'm not sure it is worth holding up the bugfix over this. The current implementation seems fine, IMO.

Reply:
I'll try to explain myself again. Imo this activity is just about the plane's perspective ("I stupidly propel forwards"), not what happens around it. If repulsion pushed the plane half across the map, should we fly all the way back? (Distance delta would get negative.) This would then become a "fly to target" activity. Imo not the goal here.

Reply:
> 
> 
> If the point is not to cover a set distance, then we shouldn't allow constructing with a distance argument in the first place.
> 
> If the point is just to fly forward for some number of ticks—even when repulsion forces may in the extreme case make the aircraft barely move at all—then just define a number of ticks and don't bother.

The distance-based constructor is just actively misleading then.